sdk/formrecognizer/azure-ai-formrecognizer/src/main/java/com/azure/ai/formrecognizer/FormRecognizerAsyncClient.java

```diff
@@ -382,18 +384,50 @@ public FormRecognizerServiceVersion getServiceVersion() {
 
     private Function<PollingContext<OperationResult>, Mono<OperationResult>> receiptStreamActivationOperation(
         Flux<ByteBuffer> data, long length, FormContentType formContentType, boolean includeTextDetails) {
-        return (pollingContext) -> {
-            try {
-                Objects.requireNonNull(data, "'data' is required and cannot be null.");
-                Objects.requireNonNull(formContentType, "'formContentType' is required and cannot be null.");
-                return service.analyzeReceiptAsyncWithResponseAsync(ContentType.fromString(formContentType.toString()),
-                    data, length, includeTextDetails)
-                    .map(response -> new OperationResult(
+        Objects.requireNonNull(data, "'data' is required and cannot be null.");
+
+        // Size of buffered data, it used to collect all data from input Flux<ByteBuffer>
+        final long[] bufferedDataSize = {0};
+        // Cached byte buffer
+        final List<ByteBuffer> cachedBuffers = new LinkedList<>();
+        // A boolean value that only used once for guessing content type on input data
+        final AtomicBoolean[] isGuessed = {new AtomicBoolean(true)};
+        // Content type on input data
+        final ContentType[] contentType = new ContentType[1];
+
+        return pollingContext -> data.filter(ByteBuffer::hasRemaining)
+            .windowUntil(buffer -> {
+
+                if (bufferedDataSize[0] >= length) {
+                    return false;
+                }
+                bufferedDataSize[0] += buffer.remaining();
+
+                // if content type is not given, do guessing the content type.
+                if (formContentType == null && isGuessed[0].get() && bufferedDataSize[0] >= 4) {
+                    byte[] bytes = buffer.array();
+                    contentType[0] = getContentType(ByteBuffer.wrap(bytes));
+                    isGuessed[0].compareAndSet(true, false);
+                }
+                // Cache the data to a list of byte buffers
+                ByteBuffer cachedBuffer = ByteBuffer.allocate(buffer.remaining()).put(buffer);
+                cachedBuffer.flip();
+                cachedBuffers.add(cachedBuffer);
+
+                if (bufferedDataSize[0] >= length) {
+                    return true;
+                }
+                return false;
+            }, true, Integer.MAX_VALUE)
+            .next()
```

Comment:
If I understand it correctly - the `data` here can be in one of the two formats: 

data = content-type + content
data = content

We've this `windowUntil` here to detect the content-type but that comes with a cost i.e. we need to allocate a new ByteBuffer (or multiple ByteBuffers if data emits byte-by-byte), we do that even if user provide us the contentType as an argument. I think when contentType is given we can call the underlying API and give it data as it is, if contentType is not given then apply the `windowUntil` path.  This way we can also simplify the logic, we don't the var  `isGuessed` or check like `formContentType == null` 

Coming to the current `windowUntil` path - I think it may not work correctly if data comes as chunks i.e.

```
data = ByteBuffer_chunk1 (containing contentType and some data),  ByteBuffer_chunk2, ByteBuffer_chunk3
```
Looks like by doing `next()` after `windowUntil`, the newData will be just ByteBuffer_chunk1. Good to double check.

You might want to apply buffer operator after windowUntil
```
.buffer(2)
.next()
.flatMap(fluxes -> {
    var contentType = getContentType(cachedBuffers);
    if (fluxes.size() == 1) {
        return callAPI(Flux.fromIterable(cachedBuffers));
    } else {
        return callAPI(Flux.fromIterable(cachedBuffers).concatWith(fluxes.get(1)));
    }
})
```

Reply:
Also with explicit contentType check

```java
if (contentType == null) {
   return callApi(data, ..);
} else {
   return data.windowUntil(..).flatMap(fluxes -> {
      var contentType = getContentType(cachedBuffers);
      if (fluxes.size() == 1) {
          return callAPI(Flux.fromIterable(cachedBuffers));
      } else {
          return callAPI(Flux.fromIterable(cachedBuffers).concatWith(fluxes.get(1)));
      }
    });
}
```

I think we just have to cache in the `windowUntil` operator i.e. no contentType check in `windowUntil`, the check can be deferred to flatmap operator.